crates/sqllib/src/dec.rs

```diff
@@ -0,0 +1,1067 @@
+use serde::ser::Error as SerError;
+use serde::{Serialize, Serializer};
+use serde_json::Number;
+use size_of::SizeOf;
+use std::{
+    fmt,
+    hash::{Hash, Hasher},
+    ops::{Add, AddAssign, Div, Mul, Neg, Sub},
+    sync::LazyLock,
+};
+
+const MAX_DIGITS: usize = 13;
+const DECIMAL_MAX_PRECISION: usize = 3 * MAX_DIGITS;
+
+pub type LargeDecimal = Decimal<MAX_DIGITS>;
+
+#[derive(Copy, Debug, Default, Clone)]
+pub struct SqlDecimal {
+    value: LargeDecimal,
+}
+
+pub(crate) type DecimalContext = Context<LargeDecimal>;
+
+#[doc(hidden)]
+pub(crate) fn is_error(context: &DecimalContext) -> bool {
+    let status = context.status();
+    status.insufficient_storage()
+        || status.overflow()
+        || status.invalid_operation()
+        || status.division_by_zero()
+        || status.division_undefined()
+        || status.division_impossible()
+        || status.division_by_zero()
+        || status.division_undefined()
+}
+
+/// Prototype clean context
+#[doc(hidden)]
+pub static CONTEXT_PROTO: LazyLock<DecimalContext> = LazyLock::new(|| {
+    let mut cx = DecimalContext::default();
+    // Default rounding mode
+    cx.set_rounding(Rounding::Down);
+    // This is 38, conveniently matching the max precision supported by severall SQL dialects
+    cx.set_max_exponent(DECIMAL_MAX_PRECISION as isize - 1)
+        .unwrap();
+    cx.set_min_exponent(-(DECIMAL_MAX_PRECISION as isize))
+        .unwrap();
+    cx
+});
+
+impl SizeOf for SqlDecimal {
+    fn size_of_children(&self, context: &mut size_of::Context) {
+        context.add(9 + 2 * MAX_DIGITS).add_distinct_allocations(1);
+    }
+}
+
+impl SqlDecimal {
+    pub fn new(value: LargeDecimal) -> Self {
+        Self { value }
+    }
+
+    pub fn get_dec(&self) -> LargeDecimal {
+        self.value
+    }
+
+    /// Return Some(self) if the context indicates no errors
+    pub fn if_ok(self, context: &DecimalContext) -> Option<Self> {
+        if is_error(context) {
+            None
+        } else {
+            Some(self)
+        }
+    }
+
+    pub fn validate(value: LargeDecimal, context: &DecimalContext, msg: &str) -> Self {
+        if is_error(context) {
+            panic!("Error during DECIMAL computation: {}", msg);
+        }
+        if value.is_zero() && value.is_negative() {
+            SqlDecimal::new(LargeDecimal::zero())
+        } else {
+            SqlDecimal::new(value)
+        }
+    }
+
+    pub fn abs(self) -> Self {
+        let mut context = CONTEXT_PROTO.clone();
+        let mut result = self.value;
+        context.abs(&mut result);
+        Self::new(result)
+    }
+
+    pub fn is_negative(self) -> bool {
+        self.value.is_negative()
+    }
+
+    pub fn parse(value: &str) -> Result<SqlDecimal, ParseDecimalError> {
+        let mut context = CONTEXT_PROTO.clone();
+        let result = context.parse(value)?;
+        Ok(Self::new(result))
+    }
+
+    pub fn from_i128_with_scale(value: i128, scale: i32) -> Self {
```

Comment:
I don't understand what this function is doing. I think you're using `rescale` to shift the number, but that's not what it does.

Reply:
This seems to work better:

```rust
    pub fn from_i128_with_scale(value: i128, scale: i32) -> Self {
        let mut context = CONTEXT_PROTO.clone();
        let mut result = context.from_i128(value);
        let scale = LargeDecimal::from(-scale);
        context.rescale(&mut result, &scale);
        context.shift(&mut result, &scale);
        Self::validate(result, &context, "from_i128_with_scale")
    }

```